REPL interpreter crashes when executing a certain sequence of statements

The REPL interpreter crashes when executing this sequence of statements:

```ruby
[1, 2].select do end
[1, 2].map do |i| i - 1a end
```

Each statement runs OK individually.

The error is:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0xa43064]

goroutine 1 [running]:
github.com/goby-lang/goby/vm.(*VM).GetREPLResult(0xc420098780, 0xc4201ac580, 0x4)
  /path/to/src/github.com/goby-lang/goby/vm/repl.go:73 +0x44
github.com/goby-lang/goby/igb.StartIgb(0xbc63a4, 0x5)
  /path/to/src/github.com/goby-lang/goby/igb/repl.go:232 +0xd34
main.main()
  /path/to/src/github.com/goby-lang/goby/goby.go:29 +0x954
```

actually I found out that this may be an issue with the Array#select method implementation

```Go
func(receiver Object, sourceLine int) builtinMethodBody {
  return func(t *thread, args []Object, blockFrame *normalCallFrame) Object {
    arr := receiver.(*ArrayObject)
    var elements []Object

    if blockFrame == nil {
          return t.vm.initErrorObject(errors.InternalError, sourceLine, errors.CantYieldWithoutBlockFormat)
        }

    // If it's an empty array, pop the block's call frame
        if len(arr.Elements) == 0 {
          t.callFrameStack.pop()
        }

    for _, obj := range arr.Elements {
          result := t.builtinMethodYield(blockFrame, obj)
          if result.Target.(*BooleanObject).value {
            elements = append(elements, obj)
          }
        }

    return t.vm.initArrayObject(elements)
  }
}
```

If the block evaluation doesn't return a boolean it causes a panic in igb next time you have an issue with an undefined variable (maybe other errors too but that's the only one i found).

```Ruby
goby ➤ ./goby -i
Goby 0.1.6 💀 🙃 😆
» [1,2].select do |x| nil end
#» nil
» 1.length
#» UndefinedMethodError: Undefined Method 'length' for 1
» case
expected next token to be WHEN, got EOF instead. Line: 0
» case
» undefined
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0xa445f4]

goroutine 1 [running]:
github.com/goby-lang/goby/vm.(*VM).GetREPLResult(0xc420090780, 0xc420165740, 0x4)
        /home/mido/dev/go/src/github.com/goby-lang/goby/vm/repl.go:73 +0x44
github.com/goby-lang/goby/igb.StartIgb(0xbc8068, 0x5)
        /home/mido/dev/go/src/github.com/goby-lang/goby/igb/repl.go:232 +0xd34
main.main()
        /home/mido/dev/go/src/github.com/goby-lang/goby/goby.go:29 +0x954
```

Still that when the vm isn't in REPL mode it actually just silently crash on method evaluation

```Ruby
goby ➤ cat samples/strange_array_select.gb

[1, 2].select do |x|
  1
end

puts("after select")
puts(undefined)
goby ➤ ./goby samples/strange_array_select.gb
goby ➤ 
goby ➤ cat samples/strange_array_select.gb

[1, 2].select do |x|
  true
end

puts("after select")
puts(undefined)
goby ➤ ./goby samples/strange_array_select.gb
after select
UndefinedMethodError: Undefined Method 'undefined' for <Instance of: Object>
from /home/mido/dev/go/src/github.com/goby-lang/goby/samples/strange_array_select.gb:7
```

I got no idea on how to handle this though ^^

--- a/vm/array.go
+++ b/vm/array.go
@@ -213,13 +213,7 @@ func builtinArrayInstanceMethods() []*BuiltinMethodObject {
 					for _, obj := range arr.Elements {
 						result := t.builtinMethodYield(blockFrame, obj)
 
-						booleanResult, isResultBoolean := result.Target.(*BooleanObject)
-
-						if isResultBoolean {
-							if booleanResult.value {
-								return TRUE
-							}
-						} else if result.Target != NULL {
+						if result.Target.isTruthy() {
 							return TRUE
 						}
 					}
@@ -327,7 +321,7 @@ func builtinArrayInstanceMethods() []*BuiltinMethodObject {
 
 						for _, obj := range arr.Elements {
 							result := t.builtinMethodYield(blockFrame, obj)
-							if result.Target.(*BooleanObject).value {
+							if result.Target.isTruthy() {
 								count++
 							}
 						}
@@ -937,7 +931,7 @@ func builtinArrayInstanceMethods() []*BuiltinMethodObject {
 
 					for _, obj := range arr.Elements {
 						result := t.builtinMethodYield(blockFrame, obj)
-						if result.Target.(*BooleanObject).value {
+						if result.Target.isTruthy() {
 							elements = append(elements, obj)
 						}
 					}

--- a/vm/boolean.go
+++ b/vm/boolean.go
@@ -143,6 +143,10 @@ func (b *BooleanObject) toJSON() string {
 	return b.toString()
 }
 
+func (b *BooleanObject) isTruthy() bool {
+	return b.value
+}
+
 // equal returns true if the Boolean values between receiver and parameter are equal
 func (b *BooleanObject) equal(e *BooleanObject) bool {
 	return b.value == e.value

--- a/vm/hash.go
+++ b/vm/hash.go
@@ -200,13 +200,7 @@ func builtinHashInstanceMethods() []*BuiltinMethodObject {
 						objectKey := t.vm.initStringObject(stringKey)
 						result := t.builtinMethodYield(blockFrame, objectKey, value)
 
-						booleanResult, isResultBoolean := result.Target.(*BooleanObject)
-
-						if isResultBoolean {
-							if booleanResult.value {
-								return TRUE
-							}
-						} else if result.Target != NULL {
+						if result.Target.isTruthy() {
 							return TRUE
 						}
 					}
@@ -937,13 +931,7 @@ func builtinHashInstanceMethods() []*BuiltinMethodObject {
 						objectKey := t.vm.initStringObject(stringKey)
 						result := t.builtinMethodYield(blockFrame, objectKey, value)
 
-						booleanResult, isResultBoolean := result.Target.(*BooleanObject)
-
-						if isResultBoolean {
-							if booleanResult.value {
-								destinationPairs[stringKey] = value
-							}
-						} else if result.Target != NULL {
+						if result.Target.isTruthy() {
 							destinationPairs[stringKey] = value
 						}
 					}

--- a/vm/null.go
+++ b/vm/null.go
@@ -159,3 +159,7 @@ func (n *NullObject) toString() string {
 func (n *NullObject) toJSON() string {
 	return "null"
 }
+
+func (n *NullObject) isTruthy() bool {
+	return false
+}

--- a/vm/object.go
+++ b/vm/object.go
@@ -17,6 +17,7 @@ type Object interface {
 	id() int
 	instanceVariableGet(string) (Object, bool)
 	instanceVariableSet(string, Object) Object
+	isTruthy() bool
 }
 
 // baseObj ==============================================================
@@ -81,6 +82,10 @@ func (b *baseObj) id() int {
 		panic(e.Error())
 	}
 	return int(r)
+}
+
+func (b *baseObj) isTruthy() bool {
+	return true
 }
 
 // Pointer ==============================================================

